### DISCUSSION TITLE
Read the contents of an uploaded file within a Flow

### DISCUSSION THREAD

**Discussion by akoenig** [directus/directus#17107] [Archive]
Currently, I want to implement an use case where an user uploads a file (within Directus) and where the uploaded file should be read, transformed and inserted into a collection. So basically some kind of ETL flow. I saw that it is possible to trigger a flow, based on the event, but not sure how to read the contents of the file afterwards.

I thought that a `Run Code` operation would do the trick, but not sure how to access the actual file. Is it required to "allow list" the `@directus/sdk` and then read the file or is there something I'm overlooking which is actually a built-in feature?

Any pointer into the right direction is highly appreciated 🙂

**Comment by paescuj** [comment]
The [run script](https://docs.directus.io/configuration/flows/operations.html#run-script) operation is rather intended to perform simpler checks or changes. For doing heavier things such as transforming files I recommend you to take a look at [custom API hooks](https://docs.directus.io/extensions/hooks.html).

**Reply by akoenig** [reply]
@paescuj, ah interesting so building an extension would be the way to go here. Thanks a lot for your rapid help, @paescuj. Highly appreciated.

By reading the docs I was wondering how the development story for extensions look like? When using [local extensions](https://docs.directus.io/extensions/creating-extensions.html#local-extension), is it required to restart the Directus instance after every change or are there any mocks to build integration tests (and develop against the mocks)?

**Reply by paescuj** [reply]
> @paescuj, ah interesting so building an extension would be the way to go here. Thanks a lot for your rapid help, @paescuj. Highly appreciated.

Yeah, that's right 👍 You're welcome!

> [...] When using [local extensions](https://docs.directus.io/extensions/creating-extensions.html#local-extension), is it required to restart the Directus instance after every change [...]

There's the `EXTENSIONS_AUTO_RELOAD` config (see https://docs.directus.io/self-hosted/config-options.html#extensions) which will automatically trigger a restart of Directus when set to `true`.

> [...] are there any mocks to build integration tests (and develop against the mocks)?

That sounds interesting 😄 But no, there's currently no such thing like mocks. But it already helps a lot if you develop the extension in TypeScript since this way you'll profit from the exposed types, see https://docs.directus.io/extensions/creating-extensions.html#creating-extensions.

**Reply by TBG-FR** [reply]
Isn't it strange that in Flows we can "Read Data" from any collection, but not from "system" collections, like files ? It feels like it should be available...

**Comment by Dvelezs94** [comment]
bumping this. I have a similar use case where I am uploading a CSV to directus files, and then I want to the read the contents of that CSV, to then update a record.

For what I read above, this should not be a super complex or heavy task, but is this possible with directus flows?

**Reply by snimavat** [reply]
@Dvelezs94 how did you endup implementing this? i am trying to do some thing similar

**Reply by jstet** [reply]
Get request to e.g.  http://127.0.0.1:8055/assets/{id} returns the data
